### ISSUE TITLE
[Feature Request] Improvements to interactive skim

### ISSUE THREAD

**Issue by EpocSquadron** [andreyorst/fzf.kak#68]
Currently if you use skim as your `fzf_implementation`, you are presented with nice, highlighted previews during a file search (`f` in fzf-mode). However, the `g` command that provides an interactive grep-with-skim option does not have this nice preview/highlighting and instead shows a single-line excerpt next to each result as you search.

We can improve upon this drastically and not only provide the preview window in the grep option, but also center the preview on the matching line, highlighted. In fact, I had achieved this functionality on startup of kakoune before I discovered your plugin. Here's what it looks like:

![image](https://user-images.githubusercontent.com/414963/64272437-5f2fbf80-cf0d-11e9-9d5c-310889ad837c.png)

Here's the shell code I used:

```bash
rkak() {
    kak $( \
        sk --ansi -i -c 'rg --hidden --line-number --color=always "{}"' --delimiter ':' --preview 'bat --style=numbers,changes --highlight-line {1} --line-range $(preview_start {1}): --color always {0}' | \
        sd '([^:]+):(\d+).*' '$1 +$2' \
    )
}
```
What's happening here is that `sk --preview` can pass parameters describing the file to be previewed as well as the line number of the match. `{1}` in this context gives us the matching line, so I pass this to `bat`'s `--highlight-line`, then tell bat to show only lines centered on that line using a helper script. You can ignore the `sd` part as that is what was formatting the filename and line number so that kakoune could open it at the matched line, which you've already figured out in your implementation.

Here's the contents of the preview_start command:

```bash
#!/bin/bash

line_to_center=`echo -n $1 | sd "'" ''`
half_height=$((`tput lines` / 2))

if (( $line_to_center > $half_height )); then
    echo -n $(( $line_to_center - $half_height ))
else
     echo -n 0
fi
```

Basically this is using `tput lines` to find the height of the window (which fzf.kak should already have knowledge of so no need to use), then integer divides by 2, subtracts that from the highlighted line number that was passed in, and returns that number. When passed to `--line-range` in `bat`, we don't need to specify the ending line number, so it gets passed as `--line-range <start_line>:`.

I don't have a very good grasp of how to write plugins for kak, nor do I have much familiarity with the workings of this plugin in particular. Perhaps with the description above you have enough to add the functionality? If not, I can try and dig in and add it to the module, but itll probably take a while.

**Comment by andreyorst** [comment]
Thanks for this request, this is nice quality of life improvement. I do not use skim myself, this module was created as part of generalization of a plugin to support compatible fuzzy frameworks, such as skim, but there's not so many of those, so I just added the missing grep feature accordingly to this feature request https://github.com/andreyorst/fzf.kak/issues/39 (extracted from https://github.com/andreyorst/fzf.kak/issues/14).

I'm going to investigate this though, since I'm interested in such feature, and maybe I'll start using skim for that.

**Comment by andreyorst** [comment]
I'm experiencing a lot of hangs and freezing when trying to get preview while using interactive grep with skim. Since I'm not active `skim` user there might be something that I do not know about the implementation and restrictions. I've tried this:

    sk -m -i -c 'grep -RHn {}' --preview '(cat $(echo {} | awk -F ":" "{ print \$1 }"))'

After executing this I can't type in the prompt and <kbd>Ctrl+C</kbd> no longer works as well. Killing `sk` from another window helps, but then terminal is in irresponsible state. I'm not sure if this feature is unstable or I have a bug, but I can't work on this feature in current state of art.

**Comment by andreyorst** [comment]
I've created a [grep-preview](https://github.com/andreyorst/fzf.kak/tree/grep-preview) branch, that adds this feature for the `fzf-grep` command. It's not exactly like skim, and probably needs more work, but please see if that suits you.

**Comment by EpocSquadron** [comment]
I've actually moved on to helix editor nowadays, but I am happy to see improvement for those still using kakoune, or in case I go back some day.

**Comment by andreyorst** [comment]
thanks for the info! I'm going to close this, as I've merged the feature for grep, and I'm going to deprecate the skim extension in the future release.
